ebos/eclgenericwriter.cc

```diff
@@ -22,6 +22,11 @@
 */
 
 #include <config.h>
+
+#ifndef HAVE_MPI
+#define HAVE_MPI 0
+#endif
```

Comment:
> This looks really strange and dangerous.

Strange indeed, but it's not dangerous (here).  This is immediately after we `#include <config.h>` so we know if `HAVE_MPI` is defined.  It's a way of suppressing warnings of the form
```
warning: "HAVE_MPI" is not defined, evaluates to 0
```
when using the `-Wundef` flag.

Reply:
Maybe we should rather fix the code that triggers the warning?

Reply:
> Maybe we should rather fix the code that triggers the warning?

Yes I think we should, but that also entails deciding how to write this check in our code base and then ensuring that the build system generates a definition that works with that way of writing it.  Right now we have a mix of
```C++
#if HAVE_MPI
```
and
```C++
#ifdef HAVE_MPI
```
with the latter mostly (or maybe exclusively) being in the opm-upscaling module.

The **safe** way to write the check in our own code is probably
```C++
#if defined(HAVE_MPI) && HAVE_MPI
```
but we also have to take third-party/upstream code into account.  Dune, at least in the 2.7.1 release I have installed on my local machine, appears to use
```C++
#if HAVE_MPI
```
which means we'll need
```C++
#define HAVE_MPI 0
```
if we don't want the warning I mentioned above.

Reply:
Looking at the config.h files we generate they either set the macro to 1 or not at all (and in Dune it seems the same). So it seems that the correct (not generating warnings) way to do the queries is with `#ifdef`, not with `#if`. Defining the macro to 1 ensures that the `#if` still gives the expected behaviour, but we should go with the warning-less style. But I hesitate: is there a good reason why `#if` is used in Dune for example? Do anyone have good guidance on this?

Reply:
You only get that warning with `-Wundef ` If you want to fix iit then this should be done via CMake passing -DHAVE_MPI=0 for serial builds, instead of cluttering the code. I don't care about the warning and doubt that many do. If you need this fixed than you might also want to providepatches to DUNE (where usually `#if HAVE_MPI` is used. Once that was recommend but it might be historic relic).

maybe for another separate PR?

Reply:
I definitely think changing this is for this PR, no. But what is best practice? Should I use #if or #ifdef? That is what I am trying to figure out.